torch/_dynamo/eval_frame.py

```diff
@@ -262,19 +264,103 @@ def innermost_fn(fn):
     return unaltered_fn
 
 
+dynamo_config_guard_ignorelist = {
+    "log_file_name",
+    "verbose",
+    "verify_correctness",  # will not affect model, will raise RuntimeError
+    # (no silent change to compilation behaviour)
+    "cache_size_limit",
+    "accumulated_cache_size_limit",
+    "print_specializations",
+    "replay_record_enabled",
+    "cprofile",  # only wraps _compile, not graph
+    "repro_after",
+    "repro_level",
+    "repro_forward_only",
+    "repro_tolerance",
+    "same_two_models_use_fp64",
+    "error_on_recompile",  # safe because: will throw error
+    "report_guard_failures",
+    "report_all_guard_failures",
+    "base_dir",  # used for minifying / logging
+    "translation_validation",
+    "translation_validation_timeout",
+    "translation_validation_no_bisect",
+    "DEBUG_DIR_VAR_NAME",
+    "debug_dir_root",
+}
+
+dynamo_config_guard_string_serialization_list = {
+    # "is_fbcode",  # function def (removed by ConfigModule visit)
+    "skipfiles_inline_module_allowlist",  # cannot pickle module (unused)
+    "constant_functions",  # PyCapsules (i.e. function ptrs)
+    "traceable_tensor_subclasses",  # Cannot pickle local object
+}
+
+# The config to restore to should dynamo compile / recompile when
+# executing from the compiled function's _TorchDynamoContext
+DYNAMO_SAVED_CONFIG: Optional[Dict[str, Any]] = None
+DYNAMO_SAVED_CONFIG_HASH: Optional[str] = None
+
+DYNAMO_CACHED_RECOMPILE_HASH: Optional[str] = None
+
+
 @contextlib.contextmanager
-def enable_dynamic(enable: Optional[bool] = None, export: bool = False):
-    if enable is None:
+def restore_compiled_dynamo_config(
+    first_ctx: bool,
+    saved_config: Dict[str, Any],
+    saved_config_hash: str,
+):
+    global DYNAMO_SAVED_CONFIG, DYNAMO_SAVED_CONFIG_HASH
+    # Set exactly once from top-level compile
+    is_top_level = False
+    try:
+        if DYNAMO_SAVED_CONFIG_HASH == DYNAMO_SAVED_CONFIG is None:
+            is_top_level = True
+            DYNAMO_SAVED_CONFIG = saved_config
+            DYNAMO_SAVED_CONFIG_HASH = saved_config_hash
+            log.debug("Set top-level compile config hash: %s", DYNAMO_SAVED_CONFIG_HASH)
+        else:
+            log.debug("Ignoring inner dynamo compile config and hash")  # TODO: remove?
         yield
-    elif enable:
-        # Assume everything is dynamic by default
-        with config.patch(assume_static_by_default=False):
-            yield
+    finally:
+        if is_top_level:
+            DYNAMO_SAVED_CONFIG_HASH = DYNAMO_SAVED_CONFIG = None
```

Comment:
not sure if this is thread safe...

Reply:
In fact, we may want to put some locks around mutating global state?

DYNAMO_SAVED_CONFIG, DYNAMO_SAVED_CONFIG_HASH, DYNAMO_CACHED_RECOMPILE_HASH?

Reply:
We could also make these vars thread-local...?

Reply:
API: `local = threading.local()`. Would you prefer this @ezyang?

Reply:
I didn't read super closely here, but entering dynamo contexts should be thread safe due to https://github.com/pytorch/torchdynamo/pull/239 - so if this is only called from there it might be okay.

Reply:
Hmm, looks like `python -m pytest test/dynamo/test_misc.py -s -vv -k test_backend_match_guard_multi_threads` is failing though - I think it's because we enter the contexts outside of the custom eval frame...

Reply:
seems to be fixed with threading.local